src/print.c

```diff
@@ -245,6 +259,189 @@ launch_preview (const char *filename,
   return retval;
 }
 
+static int
+pdf_get_actions_page (char        *filename,
+                      PDF_ACTIONS  act,
+                      int          page)
+{
+  g_autoptr (GSubprocess) process = NULL;
+  GInputStream *stream;
+  g_autoptr(GPtrArray) args = g_ptr_array_new_full (8, g_free);
+  g_autoptr (GError) err = NULL;
+  char buffer[50] = { 0 };
+
+  g_ptr_array_add (args, g_strdup_printf ("%s", LIBEXECDIR "/xdg-desktop-portal-gtk-utils/pdftoraw"));
+  g_ptr_array_add (args, g_strdup_printf ("--file=%s", filename));
+  switch (act)
+  {
+    case PDF_TEST :
+       g_ptr_array_add (args, g_strdup ("--test"));
+       break;
+    case PDF_NUM_PAGES :
+       g_ptr_array_add (args, g_strdup ("--pages"));
+       break;
+    case PDF_WIDTH :
+       g_ptr_array_add (args, g_strdup_printf ("--width=%d", page));
+       break;
+    case PDF_HEIGHT :
+       g_ptr_array_add (args, g_strdup_printf ("--height=%d", page));
+       break;
+   default:
+       return 0;
+  }
+  g_ptr_array_add (args, NULL);
+
+  process = g_subprocess_newv ((const gchar * const *)args->pdata, G_SUBPROCESS_FLAGS_STDOUT_PIPE, NULL);
+
+  if (process)
+  {
+    stream = g_subprocess_get_stdout_pipe (process);
+    if (stream)
+    {
+      g_input_stream_read (stream, buffer, sizeof (buffer), NULL, NULL);
+      if (buffer[0])
+        return atoi (buffer);
+    }
+  }
+  return 0;
+}
+
+static unsigned char *
+pdf_get_data_page (char *filename,
+                   int   w,
+                   int   h,
+                   int   page)
+{
+  g_autoptr (GSubprocess) process = NULL;
+  GInputStream *stream;
+  unsigned char *data = NULL;
+  g_autoptr (GPtrArray) args = g_ptr_array_new_full (8, g_free);
+  int read = 0;
+  gsize total_read = 0;
+  g_autoptr (GError) err = NULL;
+
+  g_ptr_array_add (args, g_strdup_printf ("%s", LIBEXECDIR "/xdg-desktop-portal-gtk-utils/pdftoraw"));
+  g_ptr_array_add (args, g_strdup_printf ("--file=%s", filename));
+  g_ptr_array_add (args, g_strdup_printf ("--raw=%d", page));
+  g_ptr_array_add (args, NULL);
+  process = g_subprocess_newv ((const gchar * const *)args->pdata, G_SUBPROCESS_FLAGS_STDOUT_PIPE, NULL);
+
+  if (process)
+  {
+    stream = g_subprocess_get_stdout_pipe (process);
+    if (stream)
+    {
+      size_t max_size = w * h * 4;
+      data = (unsigned char *) g_malloc (max_size);
+      while ((read = g_input_stream_read (stream, data + total_read, 1024, NULL, NULL)) > 0 && max_size > total_read)
```

Comment:
I don't think this logic is correct. You have to ensure the read is less than the remaining size *before* it happens.

This checks after by which time its already overran.

Reply:
(And you should read up to that exact size, not just a blind check if `1024` is enough)

Reply:
fixed